src/coreclr/vm/marshalnative.cpp

```diff
@@ -283,27 +284,45 @@ extern "C" IsInCooperativeGCMode_fn QCALLTYPE MarshalNative_GetIsInCooperativeGC
 #endif
 
 /************************************************************************
- * Marshal.GetLastPInvokeError
+ * Support for the last PInvoke error
  */
+static thread_local int t_lastPInvokeError;
+
 FCIMPL0(int, MarshalNative::GetLastPInvokeError)
 {
     FCALL_CONTRACT;
 
-    return (UINT32)(GetThread()->m_dwLastError);
+    return t_lastPInvokeError;
 }
 FCIMPLEND
 
-/************************************************************************
- * Marshal.SetLastPInvokeError
- */
 FCIMPL1(void, MarshalNative::SetLastPInvokeError, int error)
 {
     FCALL_CONTRACT;
 
-    GetThread()->m_dwLastError = (DWORD)error;
+    t_lastPInvokeError = error;
+}
+FCIMPLEND
+
+FCIMPL0(void, StubHelpers::SetLastError)
+{
+    // Make sure this is the first thing we do after returning from the target, as almost everything can cause the last error to get trashed
+    DWORD lastError = ::GetLastError();
+
+    FCALL_CONTRACT;
+
+    t_lastPInvokeError = lastError;
 }
 FCIMPLEND
 
+// GetLastError override for IJW
+DWORD STDMETHODCALLTYPE FalseGetLastError()
```

Comment:
I was about to turn this into a managed thread local, but this IJW helper method makes it a bit problematic. It can be done using UnmanagedCallersOnly, but it may come with unexpected side-effect.

Reply:
Might be worth blocking the one usage point of this API with a `TARGET_WINDOWS`. I think it might also be time to have a `FEATURE_IJW` flag. If you can create it for this we can start placing items under that in .NET 9?

Reply:
Do we expect to rename IJW in our code at some point? I assume it means NATIVEAOT.

Reply:
IJW means C++/CLI managed/native interop: https://github.com/dotnet/runtime/pull/100267/files#diff-04adba06c610a2521627fb97830440e1a70075fc02add5c949fdbc7ddbe35727R28

It is very entrenched name. We even ship a binary with this name: https://github.com/dotnet/runtime/tree/main/src/native/corehost/ijwhost